File: apps/ruby-sdk/lib/firecrawl/models/map_options.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Options for mapping (discovering URLs on) a website.
    class MapOptions
      FIELDS = %i[
        search sitemap include_subdomains ignore_query_parameters
        limit timeout integration location
      ].freeze

      attr_reader(*FIELDS)

      def initialize(**kwargs)
        FIELDS.each { |f| instance_variable_set(:"@#{f}", kwargs[f]) }
      end

      def to_h
        {
          "search" => search,
          "sitemap" => sitemap,
          "includeSubdomains" => include_subdomains,
          "ignoreQueryParameters" => ignore_query_parameters,
          "limit" => limit,
          "timeout" => timeout,
          "integration" => integration,
          "location" => location.is_a?(Hash) ? location : location&.to_h,
        }.compact
      end
    end
  end
end

File: apps/php-sdk/src/Models/MapOptions.php
<?php

declare(strict_types=1);

namespace Firecrawl\Models;

final class MapOptions
{
    private function __construct(
        private readonly ?string $search = null,
        private readonly ?string $sitemap = null,
        private readonly ?bool $includeSubdomains = null,
        private readonly ?bool $ignoreQueryParameters = null,
        private readonly ?int $limit = null,
        private readonly ?int $timeout = null,
        private readonly ?string $integration = null,
        private readonly ?LocationConfig $location = null,
    ) {}

    public static function with(
        ?string $search = null,
        ?string $sitemap = null,
        ?bool $includeSubdomains = null,
        ?bool $ignoreQueryParameters = null,
        ?int $limit = null,
        ?int $timeout = null,
        ?string $integration = null,
        ?LocationConfig $location = null,
    ): self {
        return new self(
            $search, $sitemap, $includeSubdomains, $ignoreQueryParameters,
            $limit, $timeout, $integration, $location,
        );
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        $fields = [
            'search' => $this->search,
            'sitemap' => $this->sitemap,
            'includeSubdomains' => $this->includeSubdomains,
            'ignoreQueryParameters' => $this->ignoreQueryParameters,
            'limit' => $this->limit,
            'timeout' => $this->timeout,
            'integration' => $this->integration,
            'location' => $this->location?->toArray(),
        ];

        return array_filter($fields, fn (mixed $v): bool => $v !== null);
    }
}
